### PR TITLE
Add a basic instruction as to how to use KLEE_DEBUG_WITH_TYPE

### PR DESCRIPTION
**Pull request by kren1**
I've only know learned how to use this feature. This PR briefly documents how to use it.

### CODE REVIEW

File: docs/developers-guide.md [hunk 1]
```diff
@@ -121,6 +121,25 @@ the ``klee_add_component`` function.
...
+
+KLEE provides macros that aid with tracing or "print debugging".
+
+`KLEE_DEBUG_WITH_TYPE(type, code);` which can be used as follows:
+
+```C
+#include "klee/Support/Debug.h"
+...
+KLEE_DEBUG_WITH_TYPE("my-feature",
+  llvm::errs() << "Currently there are " << states.size() << "\n";
```
Comment (251) [inline docs/developers-guide.md:134]: Maybe use `dbgs` and add " states"?
`llvm::dbgs() << "Currently there are " << states.size() << " states\n"`

**Comment by ccadar** [comment]
@kren1 Thanks, that's useful!

@251 thanks for your review.  And if you're happy with this (as well as other PRs to this repo in general), please feel free to merge it.

**Comment by MartinNowack** [comment]
@kren1 Your code example only works if you have LLVM compiled with debug information. Otherwise linking will fail.

**Comment by kren1** [comment]
How does it work currently? I don't see any difference between my example and what's currently in KLEE

**Comment by MartinNowack** [comment]
Well, if you have a debug build of KLEE but non-debug LLVM build the function might not link: http://llvm.org/doxygen/Debug_8cpp_source.html

It might just have happened with older LLVM versions.

**Comment by kren1** [comment]
Surely that's a build system problem? ie. shouldn't be allowed

**Comment by MartinNowack** [comment]
It might be not a problem anymore with newer LLVM versions. Still, the problem for me in your example is that:
```
llvm::dbgs() << "Currently there are " << states.size() << "states\n";
```
won't provide any output if LLVM is not compiled with debug enabled. 
But, the following would:
```
llvm::errs() << "Currently there are " << states.size() << "states\n";
```

**Comment by kren1** [comment]
Okay, I'll change it to `errs()`, I only changed to `dbgs()` because @251 insisted

**Comment by 251** [comment]
I did not insist. It's just strange that we use it in KLEE with `dbgs` (see main.cpp, ModuleUtil.cpp) but in the documentation we recommend a different usage. So in short - KLEE needs to be fixed?

**Comment by kren1** [comment]
There is many more uses of `errs` then the two `dbgs` you mentioned.

**Comment by 251** [comment]
I only find 10 usages of `KLEE_DEBUG_WITH_TYPE` and all use `dbgs`?!

**Comment by kren1** [comment]
I guess you are missing `KLEE_DEBUG`

**Comment by 251** [comment]
`KLEE_DEBUG` uses 4x `errs`, 1x `klee_warning`, 1x`klee_message` and 1x `.dump()`? Not sure about the "many more"...

**Comment by kren1** [comment]
Yeah I missed the usses in `tools`, but this is really besides the point. I don't see how any of this has an impact on this PR, do you want some standardisation?

**Comment by 251** [comment]
> do you want some standardisation?

No, I want proper logging for KLEE. ;) Mentioned it before several times...

For this PR I'm fine with `errs()` but maybe we should fix the `dbgs` in KLEE (@MartinNowack)?

**Comment by ccadar** [comment]
Is this really an issue?  If you look at http://llvm.org/doxygen/Debug_8cpp_source.html, when LLVM is compiled w/o debug information, `dbgs` is really the same as `errs`:

```
 #ifndef NDEBUG
...
#else
 // Avoid "has no symbols" warning.
 namespace llvm {
   /// dbgs - Return errs().
   raw_ostream &dbgs() {
     return errs();
   }
 }

#endif
```

**Comment by kren1** [comment]
I think @MartinNowack  is concerned for older LLVM versions.

**Comment by ccadar** [comment]
I think we can check older versions to see if that's an issue.  
In any case, I think we all agree this PR is fine, so let's merge it.  
We can open a separate issue if needed.

**Pull request merged**